pkg/azure/utils/concurrent.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 SAP SE or an SAP affiliate company
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"context"
+	"fmt"
+	"runtime/debug"
+	"sync"
+
+	"k8s.io/klog/v2"
+)
+
+// Task is a holder for a named function.
+type Task struct {
+	// Name is the name of the task
+	Name string
+	// Fn is the function which accepts a context and returns an error if there is one.
+	// Implementations of Fn should handle context cancellation properly.
+	Fn func(ctx context.Context) error
+}
+
+// RunGroup is a runner for concurrently spawning multiple asynchronous tasks. If any task
+// errors or panics then these are captured as errors.
+type RunGroup struct {
+	wg        sync.WaitGroup
+	semaphore chan struct{}
+	errCh     chan error
+}
+
+// NewRunGroup creates a new RunGroup.
+func NewRunGroup(numTasks, bound int) *RunGroup {
+	return &RunGroup{
+		wg:        sync.WaitGroup{},
+		semaphore: make(chan struct{}, bound),
+		errCh:     make(chan error, numTasks),
+	}
+}
+
+// RunConcurrently runs tasks concurrently with number of goroutines bounded by bound.
+// If there is a panic executing a single Task then it will capture the panic and capture it as an error
+// which will then subsequently be returned from this function. It will not propagate the panic causing the app to exit.
+func RunConcurrently(ctx context.Context, tasks []Task, bound int) []error {
+	rg := NewRunGroup(len(tasks), bound)
+	defer rg.Close()
+
+	for _, task := range tasks {
+		rg.Trigger(ctx, task)
+	}
+	return rg.WaitAndCollectErrors()
+}
+
+// Trigger executes the task in a go-routine.
+func (g *RunGroup) Trigger(ctx context.Context, task Task) {
```

Comment:
this is mostly a serial delete of nics and disks . nics serially (which is just 1) and disks serially . 
We think a better model would be the earlier approach of deleting nics and disks in parallel, but `bound` at a time. WDYT?

Reply:
I guess you got confused due to `Trigger` being an exposed method.  I have now made it un-exported. Consumers should call `RunConcurrently` which in turn calls `trigger`.

Reply:
I discussed the problem with Rishabh and i now understood the issue that you were pointing to. I got confused because your comment is at the wrong place. This is a library. It is upto the caller how these run groups are created and what gets included in one set of tasks that should run concurrently.

To make things clear, i will continue this comment on the caller side

Reply:
Ok i have now added a bound of 2. Now changed to individual delete tasks for every disk. So now if you have 1 NIC and 2 disks then a total of 3 tasks will be added to be run concurrently with a bound of 2. In other words only 2 concurrent tasks will execute at a time.